packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AzureDevOpsCredential } from './config';
+import {
+  ClientSecretCredential,
+  ManagedIdentityCredential,
+  TokenCredential,
+} from '@azure/identity';
+import {
+  AzureDevOpsCredentials,
+  AzureDevOpsCredentialsProvider,
+} from './types';
+
+type CachedAzureDevOpsCredentials = AzureDevOpsCredentials & {
+  expiresAt?: number;
+};
+
+const tenMinutes = 1000 * 60 * 10;
+
+/**
+ * A credentials provider that caches the credentials for as long as it is valid.
+ *
+ * @public
+ */
+export class CachedAzureDevOpsCredentialsProvider
+  implements AzureDevOpsCredentialsProvider
+{
+  azureDevOpsScope = '499b84ac-1321-427f-aa17-267ca6975798/.default';
+  cached: CachedAzureDevOpsCredentials | undefined;
+
+  static create(
+    credential: AzureDevOpsCredential,
+  ): CachedAzureDevOpsCredentialsProvider {
+    switch (credential.kind) {
+      case 'PersonalAccessToken':
+        return new CachedAzureDevOpsCredentialsProvider(credential.token);
+      case 'ClientSecret':
+        return new CachedAzureDevOpsCredentialsProvider(
+          new ClientSecretCredential(
+            credential.tenantId,
+            credential.clientId,
+            credential.clientSecret,
+          ),
+        );
+      case 'ManagedIdentity':
+        return new CachedAzureDevOpsCredentialsProvider(
+          new ManagedIdentityCredential(credential.clientId),
+        );
+      default:
+        // @ts-ignore
+        // eslint-disable-next-line no-case-declarations
+        const _exhaustiveCheck: never = credential;
+
+        throw new Error(
```

Comment:
Can we use `DefaultAzureCredential` as the fallback here.  This would allow the behaviour to "just work" when other components are configured with this approach.

Reply:
I refactored the `DefaultAzureDevOpsCredentialsProvider` to add a `CachedAzureDevOpsCredentialProvider` with a  `DefaultAzureCredential` when no credentials are specified for a `dev.azure.com` integration. For Azure DevOps server, the fallback remains to use no authentication.